File: internal/audioqc/fft.go
package audioqc

import (
	"math"
	"slices"
)

// fftRadix2 computes an in-place forward DFT of re/im, whose length must be a
// power of two. Iterative Cooley-Tukey; no dependencies, no allocation beyond
// the caller's slices.
func fftRadix2(re, im []float64) {
	n := len(re)
	if n <= 1 {
		return
	}

	// Bit-reversal permutation.
	for i, j := 1, 0; i < n; i++ {
		bit := n >> 1
		for ; j&bit != 0; bit >>= 1 {
			j ^= bit
		}
		j |= bit
		if i < j {
			re[i], re[j] = re[j], re[i]
			im[i], im[j] = im[j], im[i]
		}
	}

	for length := 2; length <= n; length <<= 1 {
		ang := -2 * math.Pi / float64(length)
		wRe, wIm := math.Cos(ang), math.Sin(ang)
		for i := 0; i < n; i += length {
			curRe, curIm := 1.0, 0.0
			half := length / 2
			for j := range half {
				uRe, uIm := re[i+j], im[i+j]
				vRe := re[i+j+half]*curRe - im[i+j+half]*curIm
				vIm := re[i+j+half]*curIm + im[i+j+half]*curRe
				re[i+j], im[i+j] = uRe+vRe, uIm+vIm
				re[i+j+half], im[i+j+half] = uRe-vRe, uIm-vIm
				curRe, curIm = curRe*wRe-curIm*wIm, curRe*wIm+curIm*wRe
			}
		}
	}
}

// hann returns a periodic Hann window of length n.
func hann(n int) []float64 {
	w := make([]float64, n)
	for i := range w {
		w[i] = 0.5 - 0.5*math.Cos(2*math.Pi*float64(i)/float64(n))
	}
	return w
}

// nextPow2 returns the smallest power of two >= n.
func nextPow2(n int) int {
	p := 1
	for p < n {
		p <<= 1
	}
	return p
}

// medianOf returns the median of a copy of xs. xs is not modified.
func medianOf(xs []float64) float64 {
	if len(xs) == 0 {
		return 0
	}
	c := slices.Clone(xs)
	slices.Sort(c)
	m := len(c) / 2
	if len(c)%2 == 1 {
		return c[m]
	}
	return (c[m-1] + c[m]) / 2
}

// percentile returns the p-th percentile (0..100) using nearest-rank on a copy.
func percentile(xs []float64, p float64) float64 {
	if len(xs) == 0 {
		return 0
	}
	c := slices.Clone(xs)
	slices.Sort(c)
	idx := int(math.Round(p / 100 * float64(len(c)-1)))
	idx = min(max(idx, 0), len(c)-1)
	return c[idx]
}

File: core/src/main/kotlin/mimimoto/audioqc/Fft.kt
package mimimoto.audioqc

import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.roundToInt
import kotlin.math.sin

/**
 * In-place forward DFT. [re] and [im] must have the same power-of-two length.
 * Iterative Cooley-Tukey.
 */
internal fun fftRadix2(re: DoubleArray, im: DoubleArray) {
    val n = re.size
    if (n <= 1) return

    // Bit-reversal permutation.
    var j = 0
    for (i in 1 until n) {
        var bit = n shr 1
        while (j and bit != 0) {
            j = j xor bit
            bit = bit shr 1
        }
        j = j or bit
        if (i < j) {
            re[i] = re[j].also { re[j] = re[i] }
            im[i] = im[j].also { im[j] = im[i] }
        }
    }

    var length = 2
    while (length <= n) {
        val angle = -2 * PI / length
        val wRe = cos(angle)
        val wIm = sin(angle)
        var i = 0
        while (i < n) {
            var curRe = 1.0
            var curIm = 0.0
            val half = length / 2
            for (k in 0 until half) {
                val uRe = re[i + k]
                val uIm = im[i + k]
                val vRe = re[i + k + half] * curRe - im[i + k + half] * curIm
                val vIm = re[i + k + half] * curIm + im[i + k + half] * curRe
                re[i + k] = uRe + vRe
                im[i + k] = uIm + vIm
                re[i + k + half] = uRe - vRe
                im[i + k + half] = uIm - vIm
                val nextRe = curRe * wRe - curIm * wIm
                curIm = curRe * wIm + curIm * wRe
                curRe = nextRe
            }
            i += length
        }
        length = length shl 1
    }
}

/** A periodic Hann window of length [n]. */
internal fun hann(n: Int) = DoubleArray(n) { 0.5 - 0.5 * cos(2 * PI * it / n) }

internal fun nextPow2(n: Int): Int {
    var p = 1
    while (p < n) p = p shl 1
    return p
}

/** Median of a slice, without modifying the input. */
internal fun medianOf(xs: DoubleArray, from: Int, to: Int): Double {
    val c = xs.copyOfRange(from, to)
    if (c.isEmpty()) return 0.0
    c.sort()
    val m = c.size / 2
    return if (c.size % 2 == 1) c[m] else (c[m - 1] + c[m]) / 2
}

/** Nearest-rank percentile (0..100), without modifying the input. */
internal fun percentile(xs: DoubleArray, p: Double): Double {
    if (xs.isEmpty()) return 0.0
    val c = xs.copyOf()
    c.sort()
    val idx = (p / 100 * (c.size - 1)).roundToInt().coerceIn(0, c.size - 1)
    return c[idx]
}

internal fun DoubleArray.maxAbs(): Double {
    var m = 0.0
    for (v in this) if (abs(v) > m) m = abs(v)
    return m
}
